source/common/upstream/subset_lb.cc

```diff
@@ -563,6 +563,11 @@ LocalityWeightsConstSharedPtr SubsetLoadBalancer::HostSubsetImpl::determineLocal
   if (locality_weight_aware_) {
     if (scale_locality_weight_) {
       const auto& original_hosts_per_locality = original_host_set_.hostsPerLocality().get();
+      // E.g. we can be here in static clusters with actual locality weighting before pre-init
+      // completes.
+      if (!original_host_set_.localityWeights()) {
```

Comment:
I'm a little uneasy with this as a fix, because something like static cluster is then doing an early initialization with no locality weighting and then during preinit decides to use locality weighting. Ideally all LBs are good with this, since it happens via `updateHosts`. But, if there's a cleaner approach I'm open to it. It'd probably involve some refactoring of priority state manager I think.

Reply:
Is this because we initialize the subset lb with no hosts/locality weights, then call `updateHosts` during preinit to add in hosts/weights? Would the other approach be to ensure that we're initializing the subset lb with the entire initial state instead of doing a two-phase initialization?

I would imagine it's safe if it's going through `updateHosts`, as it seems conceptually similar to locality weights being added in dynamically with EDS

Reply:
I think what happens is the static cluster is constructed, it registers hosts with https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.cc#L905 but doesn't do the full cluster priority/locality/health check setup via https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.cc#L926 until pre-init happens. When the subset LB is constructed, it sees the static cluster but isn't yet able to obtain locality weights as a result.

We could try and make sure subset LB comes up with the final state, but yeah, it's probably safe to do what's in this PR because it goes via `updateHosts`.

Reply:
Got it, thanks for explaining